doc/designs/quic-design/quic-tls.md

```diff
@@ -0,0 +1,248 @@
+QUIC-TLS Handshake Integration
+==============================
+
+QUIC reuses the TLS handshake for the establishment of keys. It does not use
+the standard TLS record layer and instead assumes responsibility for the
+confidentiality and integrity of QUIC packets itself. Only the TLS handshake is
+used. Application data is entirely protected by QUIC.
+
+QUIC_TLS Object
+---------------
+
+A QUIC-TLS handshake is managed by a QUIC_TLS object. This object provides
+3 core functions to the rest of the QUIC implementation:
+
+```c
+QUIC_TLS *ossl_quic_tls_new(SSL *s, const QUIC_TLS_ARGS *args);
+```
+
+The `ossl_quic_tls_new` function instantiates a new `QUIC_TLS` object associated
+with the QUIC Connection held in `s` and initialises it with a set of callbacks
```

Comment:
So `s` is a `QUIC_CONNECTION`-based SSL object? Is there some reason `QUIC_TLS` needs to access the entire `QUIC_CONNECTION` &mdash; wouldn't a `SSL_CONNECTION` suffice?

May as well pass it in the args struct for consistency. So basically what I see here is the CSM creating a `SSL_CONNECTION` and then a `QUIC_TLS` object which receives a pointer to the `SSL_CONNECTION`.

Reply:
> So s is a QUIC_CONNECTION-based SSL object?

No. `s` here is a pure `SSL_CONNECTION` based SSL object.

To try and explain this, let me refer you to `ossl_quic_new` which creates the "user" visible SSL object:

https://github.com/openssl/openssl/blob/1a298b00112e50718acc1fdd40b9bce482112cbf/ssl/quic/quic_impl.c#L14-L41

In the above code `qc->tls` is the `SSL_CONNECTION` based SSL object.

So, the "user" visible SSL object has a `QUIC_CONNECTION` in it. Inside that there is an "inner" SSL object that has an `SSL_CONNECTION` in it. The above `ossl_quic_new` code will actually need some tweaking to ensure the "inner" SSL object has the right SSL_METHOD associated with it (I have those tweaks on my local machine).

We need the "inner" SSL object so we can call public API calls. We need to be able to call functions such as `SSL_connect()` and it end up calling the TLS version of `SSL_connect` not the QUIC version.

I will update the document to try and explain this.

Reply:
I've updated the text to clarify this. Please take another look.